lib/SILOptimizer/Mandatory/Differentiation.h

```diff
@@ -88,6 +89,62 @@ class PostOrderPostDominanceOrder {
   }
 };
 
+/// Creates arguments in the entry block based on the function type.
+void createEntryArguments(SILFunction *f) {
+  auto *entry = f->getEntryBlock();
+  auto conv = f->getConventions();
+  auto &ctx = f->getASTContext();
+  auto moduleDecl = f->getModule().getSwiftModule();
+  assert((entry->getNumArguments() == 0 || conv.getNumSILArguments() == 0) &&
+         "Entry already has arguments?!");
+  auto createFunctionArgument = [&](SILType type) {
+    // Create a dummy parameter declaration.
+    // Necessary to prevent crash during argument explosion optimization.
+    auto loc = f->getLocation().getSourceLoc();
+    auto *decl = new (ctx)
+        ParamDecl(VarDecl::Specifier::Default, loc, loc, Identifier(), loc,
+                  Identifier(), moduleDecl);
+    decl->setType(type.getASTType());
+    entry->createFunctionArgument(type, decl);
+  };
+  for (auto indResTy : conv.getIndirectSILResultTypes())
+    createFunctionArgument(f->mapTypeIntoContext(indResTy).getAddressType());
+  for (auto paramTy : conv.getParameterSILTypes())
+    createFunctionArgument(f->mapTypeIntoContext(paramTy));
+}
+
+/// Cloner that remaps types using the target function's generic environment.
+class BasicTypeSubstCloner final
+    : public TypeSubstCloner<BasicTypeSubstCloner, SILOptFunctionBuilder> {
+
+  static SubstitutionMap getSubstitutionMap(SILFunction *target) {
+    if (auto *targetGenEnv = target->getGenericEnvironment())
+      return targetGenEnv->getForwardingSubstitutionMap();
+    return SubstitutionMap();
+  }
+
+public:
+  explicit BasicTypeSubstCloner(SILFunction *original, SILFunction *target)
+      : TypeSubstCloner(*target, *original, getSubstitutionMap(target)) {}
+
+  void postProcess(SILInstruction *orig, SILInstruction *cloned) {
+    SILClonerWithScopes::postProcess(orig, cloned);
+  }
+
+  void visit(SILInstruction *inst) {
+    TypeSubstCloner::visit(inst);
+  }
+
+  void run() {
+    auto &target = Builder.getFunction();
+    auto *entry = target.createBasicBlock();
+    createEntryArguments(&target);
```

Comment:
SILCloner already knows how to clone function arguments. I’ve been leaning towards getting rid of ‘createEntryArguments’ instead of using it in more places.

Reply:
All you need to overload in your new class is type remapping. No need to define a ‘run()’.

Reply:
My notes:
* [`SILCloner::cloneFunction`](https://github.com/apple/swift/blob/34d2524092b68a4f37693d9ef9a5591cfb6fb7e8/include/swift/SIL/SILCloner.h#L505) does clone function arguments, but is not suitable for `TypeSubstCloner` because it does not remap argument types.
    * Perhaps a good patch would be to add something like `TypeSubstCloner::cloneFunction` upstream that includes `BasicTypeSubstCloner::run()`. AFAIK currently all subclasses duplicate the same "create entry and function arguments" logic. For now, I defined `BasicTypeSubstCloner::run()` for convenience.
* Overriding [`TypeSubstCloner::postProcess`](https://github.com/apple/swift/blob/34d2524092b68a4f37693d9ef9a5591cfb6fb7e8/include/swift/SIL/TypeSubstCloner.h#L46) is necessary:
  ```
  void postProcess(SILInstruction *Orig, SILInstruction *Cloned) {
    llvm_unreachable("Clients need to explicitly call a base class impl!");
  }
  ```
* Not overriding `TypeSubstCloner::visit` is fine, thanks!

Please let me know if you have suggestions.

Reply:
Great. It’d be good to include what the original problem was and this context in the PR message.

Reply:
Updated PR description, thanks!